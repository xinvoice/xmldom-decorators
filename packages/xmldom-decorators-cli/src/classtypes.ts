import { SimpleType } from "./xsdschema";

export interface SchemaMember {
    name: string;
    javaScriptName: string;
    namespaceUri: string|null; // only attributes can have null
    type: SchemaClass; // TODO: SchemaClass?
    xmlType: 'attribute' | 'element' | 'array' | 'text';
    minOccurs?: number;
    maxOccurs?: string;
    use?: "required" | "optional" | "prohibited"
}

export interface SchemaClass {
    type: "complexType" | "simpleType" | "builtinType" | "elementType";
    javaScriptType: string;
    simpleType?: SimpleType;
    name: string;
    namespaceUri: string;
    members: SchemaMember[];
}