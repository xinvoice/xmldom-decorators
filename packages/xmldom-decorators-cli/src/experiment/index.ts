// Experiment CLI
import { Command } from 'commander';
const program = new Command();

import { version } from '../../package.json';
import { runXsdCommand } from './commands/xsd.command';

async function startExperimentCli() {
    program
        .name('xmldom-decorators-cli')
        .description('Command line tool to generate TypeScript classes from XML Schema/XSD. The TypeScript classes can be serialized and deserialized using the xmldom-decorators library.')
        .version(version)

    program.command('xsd')
        .description('Generate TypeScript classes from XML Schema/XSD')
        .argument('<xsd input file>', 'path to the XSD file')
        .option('-o, --output-dir <string>', 'output directory')
        .action((xsdPath: string, options: { outputDir?: string }) => {
            runXsdCommand(xsdPath, options);
        });

    program.parseAsync();

}

export default startExperimentCli;
