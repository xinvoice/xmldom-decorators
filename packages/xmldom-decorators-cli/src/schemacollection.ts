import * as fs from 'fs';
import { XMLDecoratorDeserializer } from "xmldom-decorators";
import { Schema, QName, SimpleType, ComplexType, Element, Attribute } from "./xsdschema";

export class SchemaCollection {
    public schemas: Schema[] = [];

    public load(fileName: string) {
        const contents = fs.readFileSync(fileName, 'utf8') as string;
        const deser = new XMLDecoratorDeserializer();
        const schema = deser.deserialize<Schema>(contents, Schema);
    
        this.schemas.push(schema);

        const lastForwardSlash = fileName.lastIndexOf("/");
        const lastBackSlash = fileName.lastIndexOf("\\");
        let lastSlash = fileName.length;
        if (lastForwardSlash !== -1 && lastBackSlash === -1) {
            lastSlash = lastForwardSlash;
        } else if (lastForwardSlash === -1 && lastBackSlash !== -1) {
            lastSlash = lastBackSlash;
        } else if (lastForwardSlash !== -1 && lastBackSlash !== -1) {
            lastSlash = Math.max(lastForwardSlash, lastBackSlash);
        }

        const relativePath = fileName.substr(0, lastSlash + 1);

        for (let include of schema.imports) {
            if (!include.schemaLocation) {
                throw new Error("Include must specify schemaLocation");
            }

            const includeFileName = relativePath + include.schemaLocation;
            this.load(includeFileName);
        }
    }

    public get root(): Schema {
        return this.schemas[0];
    }

    public getSimpleType(name: QName): SimpleType|undefined {
        for (let schema of this.schemas) {
            if ((schema.targetNamespace || "") !== name.namespaceUri) {
                continue;
            }

            for (let simpleType of schema.simpleTypes) {
                if (simpleType.name === name.localName) {
                    return simpleType;
                }
            }
        }
    }

    public getComplexType(name: QName): ComplexType|undefined {
        for (let schema of this.schemas) {
            if ((schema.targetNamespace || "") !== name.namespaceUri) {
                continue;
            }

            for (let complexType of schema.complexTypes) {
                if (complexType.name === name.localName) {
                    return complexType;
                }
            }
        }
    }

    public getElement(name: QName): Element|undefined {
        for (let schema of this.schemas) {
            if ((schema.targetNamespace || "") !== name.namespaceUri) {
                continue;
            }

            for (let element of schema.elements) {
                if (element.name === name.localName) {
                    return element;
                }
            }
        }
    }

    public getAttribute(name: QName): Attribute|undefined {
        for (let schema of this.schemas) {
            if ((schema.targetNamespace || "") !== name.namespaceUri) {
                continue;
            }

            for (let attribute of schema.attributes) {
                if (attribute.name === name.localName) {
                    return attribute;
                }
            }
        }
    }
}
