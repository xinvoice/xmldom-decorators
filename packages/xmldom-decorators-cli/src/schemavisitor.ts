import { Schema, ComplexType, SimpleType, Element, SimpleContent, ComplexContent, QName, Choice, Sequence, Attribute, AttributeGroup } from "./xsdschema";

export class SchemaVisitor {
    public visitSchema(schema: Schema) {
        for (let item of schema.items) {
            if (item instanceof Element) {
                this.visitElement(item);
            } else if (item instanceof Attribute) {
                this.visitAttribute(item);
            } else if (item instanceof AttributeGroup) {
                this.visitAttributeGroup(item);
            } else if (item instanceof ComplexType) {
                this.visitComplexType(item);
            } else if (item instanceof SimpleType) {
                this.visitSimpleType(item);
            }
        }
    }

    public visitElement(element: Element) {
        if (element.complexType) {
            this.visitComplexType(element.complexType, element);
        } else if (element.simpleType) {
            this.visitSimpleType(element.simpleType);
        }
    }

    public visitComplexType(complexType: ComplexType, element?: Element) {
        if (complexType.attributes) {
            for (let attribute of complexType.attributes) {
                this.visitAttribute(attribute);
            }
        }

        if (complexType.attributeGroups) {
            for (let attributeGroup of complexType.attributeGroups) {
                this.visitAttributeGroup(attributeGroup);
            }
        }

        if (complexType.sequence) {
            this.visitSequence(complexType.sequence);
        } else if (complexType.choice) {
            this.visitChoice(complexType.choice);
        } else if (complexType.simpleContent) {
            this.visitSimpleContent(complexType.simpleContent);
        } else if (complexType.complexContent) {
            this.visitComplexContent(complexType.complexContent);
        } else {
            // throw new Error("complexType does not define any content" + JSON.stringify(complexType, null, 2));
        }    
    }

    public visitSimpleContent(simpleContent: SimpleContent) {
        // is @XMLText 
        if (!simpleContent.extension) {
            throw new Error("Expected extension in simpleContent");
        }

        if (simpleContent.extension.attributes) {
            for (let attribute of simpleContent.extension.attributes) {
                this.visitAttribute(attribute);
            }
        }

        if (simpleContent.extension.attributeGroups) {
            for (let attribute of simpleContent.extension.attributeGroups) {
                this.visitAttributeGroup(attribute);
            }
        }
    }

    public visitExtensionBase(base: QName) {

    }

    public visitComplexContent(complexContent: ComplexContent) {
        if (complexContent.extension) {
            if (!complexContent.extension.base) {
                throw new Error("Extension base must be set");                            
            }

            this.visitExtensionBase(complexContent.extension.base);

            if (complexContent.extension.attributes) {
                for (let attribute of complexContent.extension.attributes) {
                    this.visitAttribute(attribute);
                }
            }

            if (complexContent.extension.attributeGroups) {
                for (let attributeGroup of complexContent.extension.attributeGroups) {
                    this.visitAttributeGroup(attributeGroup);
                }
            }

            if (complexContent.extension.sequence) {
                this.visitSequence(complexContent.extension.sequence);
            }

            if (complexContent.extension.choice) {
                this.visitChoice(complexContent.extension.choice);
            }

        } else if (complexContent.restriction) {
            if (!complexContent.restriction.base) {
                throw new Error("Restriction base must be set");                            
            }

            // TODO: visitRestrictionBase?
            this.visitExtensionBase(complexContent.restriction.base);

            if (complexContent.restriction.attributes) {
                for (let attribute of complexContent.restriction.attributes) {
                    this.visitAttribute(attribute);
                }
            }

            if (complexContent.restriction.attributeGroups) {
                for (let attributeGroup of complexContent.restriction.attributeGroups) {
                    this.visitAttributeGroup(attributeGroup);
                }
            }

            if (complexContent.restriction.sequence) {
                this.visitSequence(complexContent.restriction.sequence);
            }

            if (complexContent.restriction.choice) {
                this.visitChoice(complexContent.restriction.choice);
            }
        } else {
            throw new Error("complexContent must have restriction or extension");
        }
    }

    public visitSequence(sequence: Sequence) {
        if (sequence.elements) {
            for (let e of sequence.elements) {
                this.visitElement(e);
            }
        }
    }

    public visitChoice(choice: Choice) {
        if (choice.elements) {
            for (let e of choice.elements) {
                this.visitElement(e);
            }
        }
    }
    
    public visitSimpleType(simpleType: SimpleType) {
        // OK
        if (simpleType.union) {
            if (simpleType.union.simpleType) {
                if (simpleType.union.simpleType.length === 0) {
                    throw new Error("Union with simpleType must have at least one element");
                }

                this.visitSimpleType(simpleType.union.simpleType[0]);
            } else if (simpleType.union.memberTypes) {
                if (simpleType.union.memberTypes.length === 0) {
                    throw new Error("Union with memberTypes must have at least one type");
                }
                // this.visitSimpleType(simpleType.union.memberTypes[0]);
            } else {
                throw new Error("Simpletype with empty union");
            }

        }
    }

    public visitAttribute(attribute: Attribute) {
        if (attribute.simpleType) {
            this.visitSimpleType(attribute.simpleType);
        }
    }

    public visitAttributeGroup(attributeGroup: AttributeGroup) {
        if (attributeGroup.attributes) {
            for (let attribute of attributeGroup.attributes) {
                this.visitAttribute(attribute);
            }
        }
    }
}
