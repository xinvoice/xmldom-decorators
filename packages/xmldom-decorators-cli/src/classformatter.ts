import { EOL } from "os";
import { SchemaClass, SchemaMember } from "./classtypes";

export function formatClasses(classes: SchemaClass[]): string {
    const output: string[] = [];

    output.push("import { XMLRoot, XMLElement, XMLArray, XMLAttribute, XMLText } from \"@xinvoice/xmldom-decorators\";" + EOL + EOL)
    
    const reverse = new Array(...classes).reverse();

    for (let cls of reverse) {
        formatClass(cls, output);
    }

    return output.join("");
}

function formatAttributeDecorator(m: SchemaMember) {
    const options = [];

    if (m.javaScriptName !== m.name) {
        options.push("name: \"" + m.name + "\"");
    }

    if (m.namespaceUri !== null) {
        options.push("namespaceUri: \"" + m.namespaceUri + "\"");
    }

    if (options.length > 0) {
        return "{" + options.join(", ") + "}";
    }

    return "";
}

function formatElementDecorator(m: SchemaMember) {
    let options = [];

    // if (m.javaScriptName !== m.name) {
        options.push("name: \"" + m.name + "\"");
    // }

    if (m.namespaceUri !== "") {
        options.push("namespaceUri: \"" + m.namespaceUri + "\"");
    }

    if (options.length > 0) {
        return "{types: [{ " + options.join(", ")  + " }]}";
    }

    return "";
}

function formatDefault(m: SchemaMember): string {

    if (m.minOccurs == 0) {
        return "";
    }

    if (isArrayElement(m)) {
        return " = []";
    }

    if (m.type.javaScriptType === "String") {
        return " = \"\"";
    }

    if (m.type.javaScriptType === "Number") {
        return " = 0";
    }

    if (m.type.javaScriptType === "Date") {
        return " = new Date(-8640000000000000)";
    }

    if (m.type.type === "complexType") {
        return " = new " + m.type.javaScriptType + "()";
    }

    return "";
}

function isArrayElement(e: SchemaMember): boolean {
    if (e.maxOccurs === undefined) {
        return false;
    }

    if (e.maxOccurs === "unbounded") {
        return true;
    }

    const i = parseInt(e.maxOccurs);
    if (isNaN(i))
        return false;
    
    return i > 1;
}

function formatOptional(m: SchemaMember): string {
    if (m.minOccurs === 0) {
        return "?";
    }

    return "";
}

function formatArrayDecorator(m: SchemaMember) {
    const options = [];
    if (m.name !== m.javaScriptName) {
        options.push("name: \"" + m.name + "\"");
    }

    if (m.namespaceUri !== "") {
        options.push("namespaceUri: \"" + m.namespaceUri + "\"");
    }

    // "namespaceUri: \"" + m.namespaceUri + "\", itemType: () => " + m.type.javaScriptType
    options.push("itemType: () => " + m.type.javaScriptType);
    return "{" + options.join(", ") + "}";
}

const typeMap: { [key:string]: string} = {
    "Number": "number",
    "String": "string",
    "Boolean": "boolean",
    "Date": "Date",
};

function getTypeScriptType(e: SchemaClass) {
    if (e.type === "builtinType") {
        return typeMap[e.javaScriptType];
    }

    return e.javaScriptType;
}

export function formatClass(e: SchemaClass, output: string[]) {

    output.push("@XMLRoot({name: \"" + e.name + "\", namespaceUri: \"" + e.namespaceUri + "\"})" + EOL);
    output.push("export class " + e.javaScriptType + " {" + EOL);
    for (let m of e.members) {
        if (m.xmlType === "attribute") {
            output.push("    @XMLAttribute(" + formatAttributeDecorator(m) + ")" + EOL);
            output.push("    " + m.javaScriptName + formatOptional(m) + ": " + getTypeScriptType(m.type) + formatDefault(m) + ";" + EOL + EOL);
        } else if (m.xmlType === "element" && !isArrayElement(m)) {
            output.push("    @XMLElement(" + formatElementDecorator(m) + ")" + EOL);
            output.push("    " + m.javaScriptName + formatOptional(m) + ": " + getTypeScriptType(m.type) + formatDefault(m) + ";" + EOL + EOL);
        } else if (m.xmlType === "element" && isArrayElement(m)) {
            output.push("    @XMLArray({nested: false, itemTypes: [" + formatArrayDecorator(m) + "]})" + EOL);
            output.push("    " + m.javaScriptName + formatOptional(m) + ": " + getTypeScriptType(m.type) + "[]" + formatDefault(m) + ";" + EOL + EOL);
        } else if (m.xmlType === "text") {
            output.push("    @XMLText()" + EOL);
            output.push("    " + m.javaScriptName + formatOptional(m) + ": " + getTypeScriptType(m.type) + formatDefault(m) + ";" + EOL + EOL);
        } else {
            throw new Error("Unhandled xmlType " + m.xmlType);
        }
    }
    output.push("}" + EOL + EOL);
}