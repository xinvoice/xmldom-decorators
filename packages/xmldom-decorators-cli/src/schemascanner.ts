import { Element, Schema, ComplexType, QName, SimpleType, Attribute, SimpleContent } from './xsdschema';
import { SchemaCollection } from './schemacollection';
import { SchemaVisitor } from './schemavisitor';
import { SchemaClass } from './classtypes';

const builtinTypes: SchemaClass[] = [
    {
        type: "builtinType",
        name: "string",
        namespaceUri: "http://www.w3.org/2001/XMLSchema",
        javaScriptType: "String",
        members: [],
    },
    {
        type: "builtinType",
        name: "anyURI",
        namespaceUri: "http://www.w3.org/2001/XMLSchema",
        javaScriptType: "String",
        members: [],
    },
    {
        type: "builtinType",
        name: "QName",
        namespaceUri: "http://www.w3.org/2001/XMLSchema",
        javaScriptType: "String",
        members: [],
    },
    {
        type: "builtinType",
        name: "decimal",
        namespaceUri: "http://www.w3.org/2001/XMLSchema",
        javaScriptType: "Number",
        members: [],
    },
    {
        type: "builtinType",
        name: "int",
        namespaceUri: "http://www.w3.org/2001/XMLSchema",
        javaScriptType: "Number",
        members: [],
    },
    {
        type: "builtinType",
        name: "integer",
        namespaceUri: "http://www.w3.org/2001/XMLSchema",
        javaScriptType: "Number",
        members: [],
    },
    {
        type: "builtinType",
        name: "long",
        namespaceUri: "http://www.w3.org/2001/XMLSchema",
        javaScriptType: "Number",
        members: [],
    },
    {
        type: "builtinType",
        name: "float",
        namespaceUri: "http://www.w3.org/2001/XMLSchema",
        javaScriptType: "Number",
        members: [],
    },
    {
        type: "builtinType",
        name: "date",
        namespaceUri: "http://www.w3.org/2001/XMLSchema",
        javaScriptType: "Date",
        members: [],
    },
    {
        type: "builtinType",
        name: "dateTime",
        namespaceUri: "http://www.w3.org/2001/XMLSchema",
        javaScriptType: "Date",
        members: [],
    },
];

function decapitalize(str: string): string {
    return str.substr(0, 1).toLowerCase() + str.substr(1);
}

function capitalize(str: string): string {
    return str.substr(0, 1).toUpperCase() + str.substr(1);
}

function getMemberName(path: string|undefined): string|undefined {
    if (path === undefined) {
        return undefined;
    }

    return path.split(/[\s\_\-]/).map((p, index) => index == 0 ? decapitalize(p) : capitalize(p)).join("");
}

function getClassName(path: string|string[]|undefined): string|undefined {
    // capitalize first letter, split on dash/underscore, capitalize following letter

    if (path === undefined) {
        return undefined;
    }

    if (Array.isArray(path)) {
        return path.map(p => p.split(/[\s\_\-]/).map(p => capitalize(p)).join("")).join("");
    }

    return path.split(/[\s\_\-]/).map(p => capitalize(p)).join("");
}

class SchemaMapperVisitor extends SchemaVisitor {
    path: string[] = [];
    elementClasses: Map<Element, SchemaClass>;
    complexTypeClasses: Map<ComplexType, SchemaClass>;
    simpleTypeClasses: Map<SimpleType, SchemaClass>;
    schema: Schema | undefined;
    targetNamespace: string = "";

    constructor(elementClasses: Map<Element, SchemaClass>, complexTypeClasses: Map<ComplexType, SchemaClass>, simpleTypeClasses: Map<SimpleType, SchemaClass>) {
        super();

        this.elementClasses = elementClasses;
        this.complexTypeClasses = complexTypeClasses;
        this.simpleTypeClasses = simpleTypeClasses;
    }

    visitSchema(schema: Schema) {
        this.schema = schema;
        this.targetNamespace = schema.targetNamespace || "";
        super.visitSchema(schema);
        this.schema = undefined;
        this.targetNamespace = "";
    }

    visitElement(element: Element) {
        if (element.name) {
            this.path.push(element.name);
        }

        super.visitElement(element);

        if (element.name) {
            this.path.pop();
        }
    }

    visitComplexType(complexType: ComplexType, element?: Element) {

        if (complexType.name) {
            this.path.push(complexType.name);
        }

        this.complexTypeClasses.set(complexType, {
            type: "complexType",
            javaScriptType: getClassName(complexType.name) || getClassName(this.path) || "",
            name: complexType.name || "",
            namespaceUri: this.targetNamespace,
            members: []
        } );

        super.visitComplexType(complexType, element);

        if (complexType.name) {
            this.path.pop();
        }
    }
    
    visitSimpleType(simpleType: SimpleType) {
        this.simpleTypeClasses.set(simpleType, {
            type: "simpleType",
            javaScriptType: "",
            name: simpleType.name || "",
            namespaceUri: this.targetNamespace,
            simpleType: simpleType,
            members: [],
        });

        super.visitSimpleType(simpleType);
    }
}

class SchemaClassVisitor extends SchemaVisitor {
    typeStack: SchemaClass[] = [];
    mapper: SchemaMapper;
    schema: Schema | undefined;

    constructor(mapper: SchemaMapper) {
        super();

        this.mapper = mapper;
    }

    visitSchema(schema: Schema) {
        this.schema = schema;
        super.visitSchema(schema);
        this.schema = undefined;
    }

    visitComplexType(complexType: ComplexType, element?: Element) {
        const schemaClass = this.mapper.complexTypeClasses.get(complexType);
        if (!schemaClass) {
            throw new Error("Internal error");
        }

        this.typeStack.push(schemaClass);
        super.visitComplexType(complexType);
        this.typeStack.pop();
    }

    visitElement(element: Element) {
        if (this.typeStack.length == 0) {
            super.visitElement(element);
        } else {
            const top = this.typeStack[this.typeStack.length - 1];
            let elementType: SchemaClass| undefined;
            let minOccurs = element.minOccurs;
            let maxOccurs = element.maxOccurs;
            while (true) {
                if (element.type) {
                    elementType = this.mapper.getSchemaClassByQName(element.type);
                    break;
                } else if (element.complexType) {
                    elementType = this.mapper.complexTypeClasses.get(element.complexType);
                    break;
                } else if (element.simpleType) {
                    elementType = this.mapper.simpleTypeClasses.get(element.simpleType);
                    break;
                } else if (element.ref) {
                    
                    const elementRef = this.mapper.collection.getElement(element.ref);
                    if (!elementRef) {
                        throw new Error("Cannot find element reference " + element.ref.localName);
                    }

                    // TODO: reference cycle = infinite loop
                    element = elementRef;
                } else {
                    throw new Error("Unable to determine type for element " + JSON.stringify(element));
                }
            }

            // Remove restrictions on simpleType, should resolve to a builtin type
            while (elementType && elementType.simpleType && elementType.simpleType.restriction) {
                if (!elementType.simpleType.restriction.base) {
                    throw new Error("Expected base on restriction");
                }
                elementType = this.mapper.getSchemaClassByQName(elementType.simpleType.restriction.base);
            }

            if (!elementType) {
                throw new Error("Could not determine type for element " + JSON.stringify(element));
            }

            top.members.push({
                name: element.name || "",
                javaScriptName: getMemberName(element.name) || "",
                type: elementType,
                xmlType: "element",
                minOccurs: minOccurs,
                maxOccurs: maxOccurs,
                namespaceUri: this.schema ? (this.schema.targetNamespace || "") : "",
            });
            super.visitElement(element);
        }
    }

    visitAttribute(attribute: Attribute) {
        if (this.typeStack.length == 0) {
            super.visitAttribute(attribute);
        } else {
            const top = this.typeStack[this.typeStack.length - 1];
            let attributeType: SchemaClass| undefined;
            while (true) {
                if (attribute.type) {
                    attributeType = this.mapper.getSchemaClassByQName(attribute.type);
                    if (attributeType.type === "complexType") {
                        throw new Error("Attribute cannot have complex type " + attribute.type.localName + "/" + attribute.type.namespaceUri);
                    }
                    break;
                } else if (attribute.simpleType) {
                    attributeType = this.mapper.simpleTypeClasses.get(attribute.simpleType);
                    if (!attributeType) {
                        throw new Error("Could not determine simple type for attribute " + attribute.name);
                    }
                    break;
                } else if (attribute.ref) {
                    const attributeRef = this.mapper.collection.getAttribute(attribute.ref);
                    if (!attributeRef) {
                        throw new Error("Cannot find attribute reference " + attribute.ref.localName + "/" + attribute.ref.namespaceUri);
                    }

                    attribute = attributeRef;

                } else {
                    throw new Error("Unable to determine type for attribute " + JSON.stringify(attribute));
                }
            }

            // Remove restrictions on simpleType, should resolve to a builtin type
            while (attributeType && attributeType.simpleType) {

                if (attributeType.simpleType.union) {
                    // TODO: using first type in union, whether simpleType or memberTypes
                    if (attributeType.simpleType.union.simpleType) {
                        attributeType = this.mapper.simpleTypeClasses.get(attributeType.simpleType.union.simpleType[0]);
                    } else if (attributeType.simpleType.union.memberTypes) {
                        attributeType = this.mapper.getSchemaClassByQName(attributeType.simpleType.union.memberTypes[0]);
                    } else {
                        throw new Error("Expected simpleType or memberTypes in union");
                    }
                } else if (attributeType.simpleType.restriction) {
                    if (!attributeType.simpleType.restriction.base) {
                        throw new Error("Expected base on restriction");
                    }
                    attributeType = this.mapper.getSchemaClassByQName(attributeType.simpleType.restriction.base);
                } else {
                    break;
                }
            }

            if (!attributeType) {
                throw new Error("Could not determine type for attribute");
            }

            if (attribute.name  === "lang") {
                console.log(JSON.stringify(attribute))
            }

            top.members.push({
                name: attribute.name || "",
                javaScriptName: getMemberName(attribute.name) || "",
                type: attributeType,
                xmlType: "attribute",
                namespaceUri: this.schema ? (this.schema.targetNamespace || null) : null,
            });
            super.visitAttribute(attribute);
        }
    }

    visitSimpleContent(simpleContent: SimpleContent) {
        // is @XMLText 
        if (!simpleContent.extension) {
            throw new Error("Expected extension in simpleContent");
        }

        if (!simpleContent.extension.base) {
            throw new Error("Expected base in simpleContent extension");
        }

        const textType = this.mapper.getSchemaClassByQName(simpleContent.extension.base);

        const top = this.typeStack[this.typeStack.length - 1];
        top.members.push({
            name: "value",
            javaScriptName: getMemberName("value") || "",
            type: textType,
            xmlType: "text",
            namespaceUri: this.schema ? (this.schema.targetNamespace || null) : null,
        });

        super.visitSimpleContent(simpleContent);
    }
}

export class SchemaMapper {
    collection: SchemaCollection = new SchemaCollection();
    elementClasses: Map<Element, SchemaClass> = new Map();
    complexTypeClasses: Map<ComplexType, SchemaClass> = new Map();
    simpleTypeClasses: Map<SimpleType, SchemaClass> = new Map();

    public load(fileName: string) {
        this.collection.load(fileName);

        var mapper = new SchemaMapperVisitor(this.elementClasses, this.complexTypeClasses, this.simpleTypeClasses);

        for (let schema of this.collection.schemas) {
            mapper.visitSchema(schema);
        }

        var builder = new SchemaClassVisitor(this);

        for (let schema of this.collection.schemas) {
            builder.visitSchema(schema);
        }
    }

    getClasses(): SchemaClass[] {
        const rootClasses: SchemaClass[] = [];

        for (let element of this.collection.root.elements) {
            let elementType: SchemaClass|undefined;
            if (element.type) {
                elementType = this.getSchemaClassByQName(element.type);
            } else if (element.simpleType) {
                throw new Error("Root as simpleType not implemented");
            } else if (element.complexType) {
                elementType = this.complexTypeClasses.get(element.complexType);
            }

            if (!elementType) {
                throw new Error("Could not find type for element");
            }

            if (!element.name) {
                // TODO: ref?
                throw new Error("Root element must have name");
            }

            // create a new class for the root types
            rootClasses.push({
                name: element.name,
                namespaceUri: elementType.namespaceUri,
                members: elementType.members,
                type: "elementType",
                javaScriptType: element.name,
            });
        }

        const result: SchemaClass[] = [...rootClasses];

        function addClassesToResult(cls: SchemaClass) {
            for (let member of cls.members) {
                if (member.type.type === "complexType") {
                    if (result.indexOf(member.type) === -1) {
                        result.push(member.type);
                        addClassesToResult(member.type);
                    }
                }
            }
        }

        for (let rootClass of rootClasses) {
            addClassesToResult(rootClass);
        }

        return result;
    }

    public getSchemaClassByQName(name: QName): SchemaClass {
        if (name.namespaceUri === "http://www.w3.org/2001/XMLSchema") {
            return this.getBuiltinSchemaClass(name.localName);
        }
    
        let result: SchemaClass|undefined;
        const simpleType = this.collection.getSimpleType(name);
        if (simpleType) {
            result = this.simpleTypeClasses.get(simpleType);
            if (result) {
                return result;
            }
        }

        const complexType = this.collection.getComplexType(name);
        if (complexType) {
            result = this.complexTypeClasses.get(complexType); 
            if (result) {
                return result;
            }
        }

        throw new Error("Not a valid type " + JSON.stringify(name));
    }

    getBuiltinSchemaClass(localName: string): SchemaClass {
        const builtinType = builtinTypes.find(b => b.name == localName);
        if (!builtinType) {
            throw new Error("Cannot find built in type " + localName);
        }

        return builtinType;
    }
}
