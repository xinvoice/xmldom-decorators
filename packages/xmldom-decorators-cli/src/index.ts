#!/usr/bin/env node
import { SchemaMapper } from './schemascanner'
import { formatClasses } from './classformatter';
import startExperimentCli from './experiment';

// Register the experiment CLI in ./experiment/index.ts
const isExperimentUssageAllowed = process.env.EXPERIMENT === "true"

if (isExperimentUssageAllowed){
    console.warn('\x1b[33m',"WARNING: You are using the experimental cli interface.\n" +
    " Some features are not yet stable and may not work as expected.\n" 
    + "\x1b[0m");
    startExperimentCli();
    process.exit(1);
}


if (process.argv.length < 3) {
    console.log("Usage: xsd filename.xsd");
    process.exit(1);
}

const mapper = new SchemaMapper();
mapper.load(process.argv[2]);

const classes = mapper.getClasses();
console.log(formatClasses(classes));
