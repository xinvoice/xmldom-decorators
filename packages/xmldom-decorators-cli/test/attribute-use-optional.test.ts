import { SchemaMapper } from "../src/schemascanner"
import { formatClasses } from "../src/classformatter"

describe("attribute-use-optional", () => {
    it("should generate correct code for use=optional attribute", () => {
        const mapper = new SchemaMapper();
        mapper.load("test/data/optional-attribute.xsd");
        
        const classes = mapper.getClasses();
        const output = formatClasses(classes);

        expect(output).toMatchSnapshot(`
        import { XMLRoot, XMLElement, XMLArray, XMLAttribute, XMLText } from "@xinvoice/xmldom-decorators";                                      
                                                                                                                                               
      @XMLRoot({name: "exampleElement", namespaceUri: ""})                                                                                     
      export class exampleElement {                                                                                                            
          @XMLAttribute()
          optionalAttribute?: string = undefined;
      }`)
    });
});