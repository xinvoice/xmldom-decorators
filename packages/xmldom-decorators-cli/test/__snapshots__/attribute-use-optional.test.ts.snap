// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attribute-use-optional should generate correct code for use=optional attribute: 
        import { XMLRoot, XMLElement, XMLArray, XMLAttribute, XMLText } from "@xinvoice/xmldom-decorators";                                      
                                                                                                                                               
      @XMLRoot({name: "exampleElement", namespaceUri: ""})                                                                                     
      export class exampleElement {                                                                                                            
          @XMLAttribute()
          optionalAttribute?: string = undefined;
      } 1`] = `
"import { XMLRoot, XMLElement, XMLArray, XMLAttribute, XMLText } from "@xinvoice/xmldom-decorators";

@XMLRoot({name: "exampleElement", namespaceUri: ""})
export class exampleElement {
    @XMLAttribute()
    optionalAttribute?: string = undefined;

}

"
`;
